### PR TITLE
add implementation of TSD

### PR DESCRIPTION
**Pull request by ChenhongyiYang**

### CODE REVIEW

File: models/TSD/poolings.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        batch_pad = mx.sym.reshape(mx.sym.repeat(mx.sym.arange(batch_size), repeats=image_rois, axis=0), [-1,1])
+
+        fpn_roi_feats = list()
+        for stride in rcnn_stride:
+            feat_lvl = conv_fpn_feat["stride%s" % stride]
+            rois_lvl = rois_fpn["stride%s" % stride]  # [1024, 4]
+            offset_lvl = offset_fpn["stride%s" % stride] # [1024, 2, 7, 7]
+            rois_lvl = mx.sym.concat(batch_pad, rois_lvl, dim=1)
+
+            roi_feat = mx.sym.contrib.DeformablePSROIPooling(
```
Comment (huangzehao) [inline models/TSD/poolings.py:250]: Do not hard code parameters. For example, the output_dim should be the same as the number for channels of features in conv_fpn_feat.

File: models/TSD/poolings.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                group_size=1,
+                pooled_size=p.out_size,
+                part_size=0,
+                sample_per_part=4,
+                trans_std=0.1,
+                no_trans=False,
+                name='delta_r_pooled_feat_stride%s'%stride 
+            ) # [1024, 256, 7, 7]
+
+            # roi_feat = X.roi_align(
```
Comment (huangzehao) [inline models/TSD/poolings.py:265]: Remove useless codes.

File: models/TSD/poolings.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            fpn_roi_feats.append(roi_feat)
+        roi_feat = X.add_n(*fpn_roi_feats)
+
+        if p.fp16:
+            roi_feat = X.to_fp16(roi_feat, name="delta_r_roi_feat_to_fp16")
+
+        return roi_feat
+    
+    def get_roi_feature_test(self, conv_fpn_feat, rois, trans, image_rois, batch_size=1):
+        assert batch_size == 1
```
Comment (huangzehao) [inline models/TSD/poolings.py:283]: why?

File: models/TSD/poolings.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            lvl_inds_offset = F.broadcast_like(lvl_inds, lvl_offset)
+            lvl_rois = F.where(lvl_inds_roi, rois, lvl_rois)
+            lvl_offset = F.where(lvl_inds_offset, offset, lvl_offset)
+            out_rois.append(lvl_rois)
+            out_offsets.append(lvl_offset)
+            
+    return out_rois, out_offsets
+
+
+class FPNRoiAlign(RoiAlign):
```
Comment (huangzehao) [inline models/TSD/poolings.py:81]: Why copy FPNRoiAlign here?

File: models/TSD/poolings.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+import mxnet as mx
+import numpy as np
+import mxnext as X
+from utils.patch_config import patch_config_as_nothrow
+from utils.deprecated import deprecated
+
+from symbol.builder import RoiAlign, RoiExtractor
+
+
+def fpn_roi_assign(
```
Comment (huangzehao) [inline models/TSD/poolings.py:12]: Why copy fpn_roi_assign here?

File: models/TSD/metric.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+Author: Chenhongyi Yang 
+'''
+import numpy as np
+import mxnet as mx
+import os
+from core.detection_metric import EvalMetricWithSummary
+
+
+class FGAccMetric(EvalMetricWithSummary):
```
Comment (huangzehao) [inline models/TSD/metric.py:12]: Remove useless metric

File: models/TSD/bbox_head.py [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+
+    def get_reg_target(self, rois, gt_bbox):
+        '''
+        Args:
+            rois: [batch_size, image_roi, 4]
+            gt_bbox: [batch_size, max_gt_num, 4]
+        Returns:
+            reg_target: [batch_size * image_roi, num_class * 4]
+        '''
+        def get_transform(rois, gt_boxes):
```
Comment (huangzehao) [inline models/TSD/bbox_head.py:307]: Should support different bbox_encode type such as xyxy

File: models/TSD/bbox_head.py [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+            name='tsd_bbox_reg_l1'
+        )
+        tsd_reg_loss = bbox_weight * tsd_reg_loss
+        tsd_reg_loss = X.loss(
+            tsd_reg_loss,
+            grad_scale=1.0 / batch_roi * scale_loss_shift,
+            name='tsd_bbox_reg_loss',
+        )  
+        tsd_cls_pc_loss = self.cls_pc_loss(cls_logit, tsd_cls_logit, cls_label, scale_loss_shift)
+        tsd_reg_pc_loss = self.reg_pc_loss(bbox_delta, tsd_bbox_delta, rois, rois_r, gt_bbox, cls_label, scale_loss_shift)
```
Comment (huangzehao) [inline models/TSD/bbox_head.py:241]: Skip the computation of pc_loss based on config.

File: config/TSD/tsd_r50_rpn_1x.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+
+    class BboxParam:
+        fp16 = General.fp16
+        normalizer = NormalizeParam.normalizer
+        num_class  = 1 + 80
+        image_roi  = 512
+        # -----------------------------------------
+        # info used by TSD
+        batch_size = General.batch_image
```
Comment (huangzehao) [inline config/TSD/tsd_r50_rpn_1x.py:99]: remove batch_size and use batch_image instead

File: config/TSD/tsd_r50_rpn_1x.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        roi_size   = 7 
+        in_channels = 256
+        rpn_params = RpnParam
```
Comment (huangzehao) [inline config/TSD/tsd_r50_rpn_1x.py:102]: useless param

File: config/TSD/tsd_r50_rpn_1x.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+    class BboxParam:
+        fp16 = General.fp16
+        normalizer = NormalizeParam.normalizer
+        num_class  = 1 + 80
+        image_roi  = 512
+        # -----------------------------------------
+        # info used by TSD
+        batch_size = General.batch_image
+        roi_size   = 7 
+        in_channels = 256
```
Comment (huangzehao) [inline config/TSD/tsd_r50_rpn_1x.py:101]: useless param

File: models/TSD/bbox_head.py [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+import mxnet as mx
+import mxnext as X
+from symbol.builder import BboxHead
+
+from models.TSD.poolings import FPNRoIAlign_DeltaC, FPNRoIAlign_DeltaR
+
+from shape_tool import infer_shape
+
+
+def get_iou_mat(rois, gt_boxes, l1, l2):
```
Comment (huangzehao) [inline models/TSD/bbox_head.py:12]: Can we implement this function without knowing l2?
